File: Easy/Exit Point in a Matrix/exit-point-in-a-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    pair<int, int> dir(string d) {
        if(d == "left")  return make_pair(0, -1);
        if(d == "right") return make_pair(0, 1);
        if(d == "up")    return make_pair(-1, 0);
        return make_pair(1, 0); // if(d == "down")
    }
    
    string moveRight(string d) {
        if(d == "left")  return "up";
        if(d == "right") return "down";
        if(d == "up")    return "right";
        return "left"; // if(d == "down")
    }
    
    vector<int> FindExitPoint(int n, int m, vector<vector<int>>& mat) {
        string currDir = "right";
        vector<int> ans(2);
        int i = 0, j = 0;
        
        do {
            ans[0] = i; ans[1] = j;
            
            if(mat[i][j] == 1) {
                mat[i][j] = 0;
                currDir = moveRight(currDir);
            }
            
            i += dir(currDir).first;
            j += dir(currDir).second;
            
        } while(i >= 0 && i < mat.size() && j >= 0 && j < mat[0].size());
        
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> matrix(n, vector<int>(m, 0));
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                cin >> matrix[i][j];
        Solution obj;
        vector<int> ans = obj.FindExitPoint(n, m, matrix);
        for (auto i : ans)
            cout << i << " ";
        cout << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Exit Point in a Matrix/exit-point-in-a-matrix.java
class Solution {
    public List<Integer> exitPoint(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        List<Integer> ans = new ArrayList<>();
        int i = 0, j = 0, dir = 0;
        while(i < n && j < m) {
            if(mat[i][j] == 1) {
                dir = (dir + 1) % 4;
                mat[i][j] = 0;
            }
            int ni = i + dirs[dir][0];
            int nj = j + dirs[dir][1];
            if(ni < 0 || ni >= n || nj < 0 || nj >= m) {
                ans.add(i);
                ans.add(j);
                break;
            }
            i = ni;
            j = nj;
        }
        return ans;
    }
}